Application/ControlMsrE2/ControlMsrE2UI.c

```diff
@@ -0,0 +1,277 @@
+    
+    int pos = 0;
+    STATIC CHAR16  Output[] = L"A";
+    
+    gST->ConOut->EnableCursor(gST->ConOut, TRUE);
+    
+    INT32 startRow = gST->ConOut->Mode->CursorRow;
+    INT32 startColumn = gST->ConOut->Mode->CursorColumn;
+    
+    do {
+        //
+        // Read a key
+        //
+        gBS->WaitForEvent (1, &gST->ConIn->WaitForKey, &EventIndex);
+        ZeroMem (&Key, sizeof (Key));
+        Status = gST->ConIn->ReadKeyStroke (gST->ConIn, &Key);
+        if (EFI_ERROR (Status)) {
+            
+            if (Status == EFI_NOT_READY) {
+                continue;
+            }
+            
+            break;
+        }
+        
+        switch (Key.UnicodeChar) {
+            case CHAR_BACKSPACE:
+                if (pos > 0) {
+                    pos--;
+                    gST->ConOut->SetCursorPosition (gST->ConOut, startColumn + pos, startRow);
+                    gST->ConOut->OutputString (gST->ConOut, L" ");
+                    gST->ConOut->SetCursorPosition (gST->ConOut, startColumn + pos, startRow);
+                }
+                break;
+            case 0x1B:
+                if (pos > 0) {
+                    pos = 0;
+                    gST->ConOut->SetCursorPosition (gST->ConOut, startColumn + pos, startRow);
+                    for (int i = 1; i < length; i++) {
+                        gST->ConOut->OutputString (gST->ConOut, L" ");
+                    }
+                    gST->ConOut->SetCursorPosition (gST->ConOut, startColumn + pos, startRow);
+                }
+                else {
+                    buffer[pos] = 0;
+                    return pos;
+                }
+                break;
+            case CHAR_CARRIAGE_RETURN:
+            case CHAR_LINEFEED:
+                buffer[pos] = 0;
+                gST->ConOut->EnableCursor(gST->ConOut, FALSE);
+                return pos;
+            default:
+                buffer[pos++] = Key.UnicodeChar;
+                Output[0] = Key.UnicodeChar;
+                gST->ConOut->OutputString (gST->ConOut, Output);
+                
+                if (pos >= length - 1) {
+                    buffer[pos] = 0;
+                    return pos;
+                }
+                break;
+        }
+    } while (TRUE);
+    
+    return 0;
+}
+
+CHAR16 ReadAnyKey () {
+    CHAR16 keys[2];
+    ReadLine(keys, 2);
+    return keys[0];
+}
+
+UINT32 ReadYN () {
+    CHAR16 keys[2];
+    do {
+        ReadLine(keys, 2);
+    } while (!StrContains (L"yn", keys));
+    
+    return keys[0] == 'y' || keys[0] == 'Y';
+}
+
+VOID PrintGuid (IN EFI_GUID* Guid) {
+    
+    Print (L"%08x-%04x-%04x-%02x-%02x-%02x-%02x-%02x-%02x-%02x-%02x",
+           Guid->Data1,
+           Guid->Data2,
+           Guid->Data3,
+           Guid->Data4[0], Guid->Data4[1], Guid->Data4[2], Guid->Data4[3], Guid->Data4[4], Guid->Data4[5], Guid->Data4[6], Guid->Data4[7]
+           );
+}
+
+VOID PrintUINT8Str (UINT8 n[1]) {
+    AsciiPrint ((CHAR8*)(&n[0]));
```

Comment:
This looks really ugly and unsafe, cannot you use `AsciiPrint ("%a", (CHAR8 *) X);`? Perhaps even without this function.

Reply:
I can if you insist, but it's a convenience method and IMHO PrintUInt8Str (whatever) look nicer than AsciiPrint ("%a", (CHAR8 *) whatever);